imp: add query and sudo message types to encapsulate all variants of calls

crodriguezvega opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

See item 6.7 of report (TODO: add link when report is available in ibc-go).

Using [the code](https://github.com/CosmWasm/wasmvm/blob/1638725b25d799f078d053391945399cb35664b1/types/msg.go#L248-L273) provided by Ethan as reference.

There are parts of this PR that are incomplete because changes from #4033 and #4034 are needed.

closes: #XXXX


### Commit Message / Changelog Entry

```bash
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)


<!--
Example commit messages:

fix: skip emission of unpopulated memo field in ics20
deps: updating sdk to v0.46.4
chore: removed unused variables
e2e: adding e2e upgrade test for ibc-go/v6
docs: ics27 v6 documentation updates
feat: add semantic version utilities for e2e tests
feat(api)!: this is an api breaking feature
fix(statemachine)!: this is a statemachine breaking fix
-->

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

crodriguezvega commented on modules/light-clients/08-wasm/types/msgs.go:
```diff
@@ -33,3 +35,73 @@ func (m MsgStoreCode) GetSigners() []sdk.AccAddress {
+		ConsensusState *ConsensusState `json:"consensus_state"`
+	}
+	updateStateMsg struct {
+		Header *Header `json:"header,omitempty"`
```
I changed this from the [original message](https://github.com/cosmos/ibc-go/blob/feat/wasm-clients/modules/light-clients/08-wasm/types/update.go#L53-L55), since in theory for update state you only use the header and not misbehaviour. But this change also requires changes in the contracts, so I need to check with them.

crodriguezvega commented on modules/light-clients/08-wasm/types/msgs.go:
```diff
@@ -33,3 +35,73 @@ func (m MsgStoreCode) GetSigners() []sdk.AccAddress {
+	CheckForMisbehaviour *checkForMisbehaviourMsg `json:"check_for_misbehaviour,omitempty"`
+}
+
+type (
```
I will add some godoc for these types.

crodriguezvega replied:
Done. No more block type definitions. 🕺

crodriguezvega commented on modules/light-clients/08-wasm/types/msgs.go:
```diff
@@ -33,3 +35,70 @@ func (m MsgStoreCode) GetSigners() []sdk.AccAddress {
+		Path             exported.Path   `json:"path"`
+	}
+	checkForMisbehaviourMsg struct {
+		ClientMessage exported.ClientMessage `json:"client_message"`
```
Instead of having two fields (`Header` and `Misbehaviour`) we pass directly `ClientMessage`, as we discussed during our own internal walkthrough. I think I mentioned this already to Composable, but I will double check.

crodriguezvega commented on proto/ibc/lightclients/wasm/v1/wasm.proto:
```diff
@@ -43,3 +43,10 @@ message Misbehaviour {
 
   bytes data = 1;
 }
+
```
I brought this over from #3958.

DimitrisJim commented on modules/light-clients/08-wasm/types/msgs.go:
```diff
@@ -33,3 +35,70 @@ func (m MsgStoreCode) GetSigners() []sdk.AccAddress {
 	return []sdk.AccAddress{signer}
 }
+
+type QueryMsg struct {
```
Wondering if these might be best suited in `vm`?. `msgs.go` was inspired by wasmvm's placement yeah? Though their `types/msgs.go` doesn't really have the typical interpretation as found in modules (implementing sdk.Msg and all).

damiannolan replied:
Agree on highlighting this! Would be nice to find a nice home for these specific messages. I know `msgs.go` is usually used for implementations of `sdk.Msg` (anything that gets routed through a grpc service). 

But these messages are kind of special. I don't have a problem keeping them in `msgs.go` personally, but I could also see value in some other dedicated file like `api.go` or similar which signals their intent. Also having them separate is probably a good call for maintainabiilty/visibilty, as these msgs cannot really change once the API is published

crodriguezvega replied:
I moved them to a file called `contract_api.go`. I hope the filename is fine for you!

damiannolan commented on modules/light-clients/08-wasm/types/msgs.go:
```diff
@@ -33,3 +35,70 @@ func (m MsgStoreCode) GetSigners() []sdk.AccAddress {
 	return []sdk.AccAddress{signer}
 }
+
+type QueryMsg struct {
```
we should probably add in the godoc that `omitempty` is crucial for json serialization to be compatible with rust contract enum matching

crodriguezvega replied:
Added godoc and also realised that the types could be unexported.

damiannolan reviewed (commented):
Nice, I really like the `QueryMsg` and `SudoMsg` types! Much cleaner than having all of these inner/outer things spread around! 

I left a couple of comments/suggestions, again nothing major!

damiannolan commented on modules/light-clients/08-wasm/types/errors.go:
```diff
@@ -3,13 +3,14 @@ package types
-	ErrInvalid         = errorsmod.Register(ModuleName, 1, "invalid")
-	ErrInvalidData     = errorsmod.Register(ModuleName, 2, "invalid data")
-	ErrInvalidCodeHash = errorsmod.Register(ModuleName, 3, "invalid code hash")
+	ErrInvalid              = errorsmod.Register(ModuleName, 1, "invalid")
```
Tbh I'm not sure how much truth there is to this anymore, if starting at 2 is required or not!

crodriguezvega replied:
I can change this anyway.

crodriguezvega commented on modules/light-clients/08-wasm/types/msgs.go:
```diff
@@ -33,3 +35,70 @@ func (m MsgStoreCode) GetSigners() []sdk.AccAddress {
+	}
+)
+
+type SudoMsg struct {
```
Added.

crodriguezvega commented on modules/light-clients/08-wasm/types/contract_api.go:
```diff
@@ -0,0 +1,121 @@
+	CheckForMisbehaviour *checkForMisbehaviourMsg `json:"check_for_misbehaviour,omitempty"`
+}
+
+type statusMsg struct{}
+type exportMetadataMsg struct{}
+type timestampAtHeightMsg struct {
+	Height exported.Height `json:"height"`
+}
```
gofumpt disagrees with you guys, @damiannolan and @chatton, and wants to put here a type block. I respect your opinion a lot, guys, you know that, but my religion doesn't allow me to break the 11th commandment: "You will never fight against your beloved linter". :)

crodriguezvega commented on modules/light-clients/08-wasm/types/update.go:
```diff
@@ -11,66 +13,32 @@ import (
-		ClientMessage: clientMsgConcrete,
+	clientMessage, ok := clientMsg.(*ClientMessage)
+	if !ok {
+		return errorsmod.Wrapf(ErrInvalidClientMessage, "expected type %T, got %T", &ClientMessage{}, clientMsg)
```
I think this is ok because of the casting `clientMessage, ok := clientMsg.(*ClientMessage)` above?

crodriguezvega commented:
I addressed the comments and, even though I don't have the required blessings, I will go ahead and merge it. We can fix anything else in follow-up PRs.
